### ISSUE TITLE
LTO build with MinGW on Windows produces binary dependent on build environment

### ISSUE THREAD

**Issue by niklasf** [official-stockfish/Stockfish#3990]
Stockfish 15 compiled with GitHub Actions on Windows using MinGW fails to run when downloaded. Immediately after starting, it exits without any error message or output.

It runs successfully in the build environment and can complete a profile-build.

WinDbg session after downloading to Windows 11 machine:

```
************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00000000`00400000 00000000`0065c000   image00000000`00400000
ModLoad: 00007ffa`13980000 00007ffa`13b89000   ntdll.dll
ModLoad: 00007ffa`13200000 00007ffa`132bd000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007ffa`11330000 00007ffa`116a4000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007ffa`11e20000 00007ffa`11ece000   C:\WINDOWS\System32\ADVAPI32.dll
ModLoad: 00007ffa`11890000 00007ffa`11933000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00007ffa`12c10000 00007ffa`12cae000   C:\WINDOWS\System32\sechost.dll
ModLoad: 00007ffa`12880000 00007ffa`129a0000   C:\WINDOWS\System32\RPCRT4.dll
(4e38.3934): Break instruction exception - code 80000003 (first chance)
ntdll!LdrpDoDebuggerBreak+0x30:
00007ffa`13a5cc74 cc              int     3
0:000> g
(4e38.3934): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
```

Related to #2769, but `g++ --version`:

```
g++.exe (Rev10, Built by MSYS2 project) 11.2.0
```

Running on `windows-2019` or `windows-2022`.

Full workflow run: https://github.com/lichess-org/fishnet/runs/6081765372, downstream issue https://github.com/lichess-org/fishnet/issues/227.

https://github.com/niklasf/Stockfish/commit/4fa5daf44095b164e3759e671967e6f769f5a632 works around successfully.

**Comment by vondele** [comment]
somehow a bug in the toolchain I would expect. Does adding -static to the build help? 

there is a bug tracker at https://sourceforge.net/projects/mingw-w64/ but I guess we're not specific enough yet?

**Comment by niklasf** [comment]
https://sourceforge.net/p/mingw-w64/bugs/557/ (assumed to be fixed, but then reopened) seems interesting, although the symptoms are slightly different.

The flags
```
[fishnet 2.6.1] CXXFLAGS: [...] -Wall -Wcast-qual -fno-exceptions -std=c++17  -pedantic -Wextra -Wshadow -DNDEBUG -O3 -DIS_64BIT -msse -msse3 -mpopcnt -DUSE_POPCNT -DUSE_AVX2 -mavx2 -DUSE_AVX512 -mavx512f -mavx512bw -DUSE_VNNI -mavx512vnni -mavx512dq -mavx512vl -DUSE_SSE41 -msse4.1 -DUSE_SSSE3 -mssse3 -DUSE_SSE2 -msse2 -DUSE_PEXT -mbmi2 -flto
[fishnet 2.6.1] LDFLAGS:  -static [...] -Wall -Wcast-qual -fno-exceptions -std=c++17  -pedantic -Wextra -Wshadow -DNDEBUG -O3 -DIS_64BIT -msse -msse3 -mpopcnt -DUSE_POPCNT -DUSE_AVX2 -mavx2 -DUSE_AVX512 -mavx512f -mavx512bw -DUSE_VNNI -mavx512vnni -mavx512dq -mavx512vl -DUSE_SSE41 -msse4.1 -DUSE_SSSE3 -mssse3 -DUSE_SSE2 -msse2 -DUSE_PEXT -mbmi2 -flto -save-temps
```
already include `-static` due to https://github.com/official-stockfish/Stockfish/blob/6e0680efa0d0653686b8eb8753dc49718b95cb2b/src/Makefile#L380-L382

**Comment by ppigazzini** [comment]
I'm a bit confused.

This PR https://github.com/official-stockfish/Stockfish/pull/3891 improved the Makefile for windows builds:
- `-static` is set checking `OS` for a windows host and not with `COMP` . Before the PR the SF GitHub Actions probably made a dynamic linked build)
- LTO works fine with a recent mingw-w64 (as the one provided by msys2)

The `-flto=auto` was dropped some days after that PR by:
https://github.com/official-stockfish/Stockfish/pull/3898

The Stockfish GitHub Actions logs seems to be fines.

**Comment by ppigazzini** [comment]
@niklasf I just peeked your workflow run https://github.com/lichess-org/fishnet/runs/6081765372

Install msys2 `make`:
```
Installing additional packages through pacman...
  C:\Windows\system32\cmd.exe /D /S /C D:\a\_temp\setup-msys2\msys2.cmd -c "'pacman' '--noconfirm' '-S' '--needed' '--overwrite' '*' 'mingw-w64-x86_64-gcc' 'make' 'tar'"
  resolving dependencies...
```
But use Chocolatey `mingw32-make`:
```
[fishnet 2.6.2-dev] C:/ProgramData/Chocolatey/lib/mingw/tools/install/mingw64/bin/mingw32-make ARCH=x86-64-vnni512 COMP=mingw gcc-profile-make
[fishnet 2.6.2-dev] mingw32-make[1]: Entering directory 'D:/a/fishnet/fishnet/Stockfish/src'
[fishnet 2.6.2-dev] C:/ProgramData/Chocolatey/lib/mingw/tools/install/mingw64/bin/mingw32-make ARCH=x86-64-vnni512 COMP=mingw \ 
```

**Comment by niklasf** [comment]
Oh! I'll try to make a reduced workflow run to narrow down the issue.

**Comment by niklasf** [comment]
Yep, that's it, sorry! It looks like the actual build happens in a different shell than `g++ --version`, and is likely using an older buggy toolchain.

This only happens in the lichess-org/fishnet workflow. The official CI does not have the same issue. To verify, I [modified](https://github.com/niklasf/Stockfish/commit/ca5f9da7e5e528010d63a77fa4beb6950f8863ee) the offficial workflow run to upload the binary (and removed a few extra steps): https://github.com/niklasf/Stockfish/actions/runs/2195225495. The produced binary works fine after downloading.

**Comment by ppigazzini** [comment]
I think that the workflow was using an old version of https://community.chocolatey.org/packages/mingw , the log shown `GNU Make 4.2.1`
A fresh install of that packages has:
```
mingw32-make -v
GNU Make 4.3

g++ -v
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (MinGW-W64 x86_64-posix-seh, built by Brecht Sanders)
```

**Comment by niklasf** [comment]
That's exactly what was happening. The `windows-2019` image comes with the old toolchain, and I installed the latest toolchain using `msys2/setup-msys2@v2`.

Thanks to

```
defaults:
  run:
    shell: msys2 {0}
```

commands like `run: g++ --version` would get the new toolchain, but commands issued by the Rust build script running using `actions-rs/cargo@v1` were using the old toolchain.

Fixed by installing Rust from msys2 instead of using `actions-rs`.
